internal/fourslash/fourslash.go

```diff
@@ -0,0 +1,478 @@
+package fourslash
+
+import (
+	"fmt"
+	"io"
+	"strings"
+	"testing"
+	"unicode"
+	"unicode/utf8"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/microsoft/typescript-go/internal/ast"
+	"github.com/microsoft/typescript-go/internal/bundled"
+	"github.com/microsoft/typescript-go/internal/collections"
+	"github.com/microsoft/typescript-go/internal/core"
+	"github.com/microsoft/typescript-go/internal/ls"
+	"github.com/microsoft/typescript-go/internal/lsp"
+	"github.com/microsoft/typescript-go/internal/lsp/lsproto"
+	"github.com/microsoft/typescript-go/internal/testutil/harnessutil"
+	"github.com/microsoft/typescript-go/internal/tspath"
+	"github.com/microsoft/typescript-go/internal/vfs/vfstest"
+	"gotest.tools/v3/assert"
+)
+
+type FourslashTest struct {
+	server *lsp.Server
+	in     *lspWriter
+	out    *lspReader
+	id     int32
+
+	testData *TestData
+
+	currentCaretPosition lsproto.Position
+	currentFilename      string
+	lastKnownMarkerName  string
+	activeFilename       string
+}
+
+type lspReader struct {
+	c <-chan *lsproto.Message
```

Comment:
I feel suspicious of the concurrency in this, solely because it will become normal for the server to make calls back into the client (which can then make calls back into the server, ad infinitum).

Channels are the right thing of course, maybe this is just something to consider when writing the harness itself, which sends off a request and then is presumably sitting there waiting for the reply to the request.